File: cases/taint/python/function-field-positive/function_field.py
def dfb_source():  # DFB-SOURCE: function-field-input
    return "tainted"


def dfb_sink(value):  # DFB-SINK: function-field-sink
    pass


class Holder:
    def __init__(self):
        self.fn = None


def leak(value):
    dfb_sink(value)


def drop(value):
    dfb_sink("clean")


def dispatch(holder, value):
    holder.fn(value)


def run():
    holder = Holder()
    holder.fn = leak  # DFB-WITNESS: function-field-store
    other = Holder()
    other.fn = drop
    dispatch(holder, dfb_source())

File: cases/taint/ruby/function-field-positive/function_field.rb
class Holder
  attr_accessor :fn

  def initialize
    @fn = nil
  end
end

def dfb_source # DFB-SOURCE: function-field-input
  "tainted"
end

def dfb_sink(value) # DFB-SINK: function-field-sink
end

def leak(value)
  dfb_sink(value)
end

def drop(value)
  dfb_sink("clean")
end

def dispatch(holder, value)
  holder.fn.call(value)
end

def run
  holder = Holder.new
  holder.fn = method(:leak) # DFB-WITNESS: function-field-store
  other = Holder.new
  other.fn = method(:drop)
  dispatch(holder, dfb_source)
end
